aicsimageio/readers/reader.py

```diff
@@ -0,0 +1,393 @@
+
+        Slice selection
+
+        >>> img = Reader("s_1_t_1_c_10_z_20.ome.tiff")
+        ... every_other = img.get_image_dask_data("CZYX", C=slice(0, -1, 2))
+
+        Notes
+        -----
+        * If a requested dimension is not present in the data the dimension is
+          added with a depth of 1.
+
+        See `aicsimageio.transforms.reshape_data` for more details.
+        """
+        pass
+
+    def get_image_data(
+        self, out_orientation: Optional[str] = None, **kwargs
+    ) -> da.Array:
+        """
+        Get specific dimension image data out of an image as a numpy array.
+
+        Parameters
+        ----------
+        out_orientation: Optional[str]
+            A string containing the dimension ordering desired for the returned ndarray.
+            Default: The natural image dimension order.
+
+        kwargs: Any
+            * C=1: specifies Channel 1
+            * T=3: specifies the fourth index in T
+            * D=n: D is Dimension letter and n is the index desired. D should not be
+              present in the out_orientation.
+            * D=[a, b, c]: D is Dimension letter and a, b, c is the list of indices
+              desired. D should be present in the out_orientation.
+            * D=(a, b, c): D is Dimension letter and a, b, c is the tuple of indices
+              desired. D should be present in the out_orientation.
+            * D=range(...): D is Dimension letter and range is the standard Python
+              range function. D should be present in the out_orientation.
+            * D=slice(...): D is Dimension letter and slice is the standard Python
+              slice function. D should be present in the out_orientation.
+
+        Returns
+        -------
+        data: numpy array
+            The image data with the specified dimension ordering.
+
+        Examples
+        --------
+        Specific index selection
+
+        >>> img = Reader("s_1_t_1_c_10_z_20.ome.tiff")
+        ... c1 = img.get_image_data("ZYX", C=1)
+
+        List of index selection
+
+        >>> img = Reader("s_1_t_1_c_10_z_20.ome.tiff")
+        ... first_and_second = img.get_image_data("CZYX", C=[0, 1])
+
+        Tuple of index selection
+
+        >>> img = Reader("s_1_t_1_c_10_z_20.ome.tiff")
+        ... first_and_last = img.get_image_data("CZYX", C=(0, -1))
+
+        Range of index selection
+
+        >>> img = Reader("s_1_t_1_c_10_z_20.ome.tiff")
+        ... first_three = img.get_image_dask_data("CZYX", C=range(3))
+
+        Slice selection
+
+        >>> img = Reader("s_1_t_1_c_10_z_20.ome.tiff")
+        ... every_other = img.get_image_data("CZYX", C=slice(0, -1, 2))
+
+        Notes
+        -----
+        * If a requested dimension is not present in the data the dimension is
+          added with a depth of 1.
+
+        See `aicsimageio.transforms.reshape_data` for more details.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def metadata(self) -> Any:
```

Comment:
Do we want to specify a return type. Currently, you'd have to probe and then customize your code for the readers behavior you happen to be using but this would produce brittle code.

Reply:
No we have always said it will be `Any` because it depends on the file format as to which metadata object type makes sense to return. `DefaultReader` should return a Dict of EXIF info imo.

Reply:
There is a world where we always convert to ome-xml but if I'm using a specific reader and I know it, then I should be able to get the native metadata format also.   This isn't really about polymorphism at this point, it's about just having a commonly named function.

Reply:
I think for the AICSImage object I can see always emulating OMEXML would be awesome but for the underlying readers it should definitely be the base format.